### ISSUE TITLE
Altering system_distributed and system_traces does not result in an error

### ISSUE THREAD

**Issue by ShlomiBalalis** [scylladb/scylla#7629]
Scylla version: 4.3
Commit https://github.com/scylladb/scylla/commit/0143aaa5a8f180b76dbb0dbdc50844bba2bf0899 was suppose to stop users from altering the internal (system) keyspaces.

When trying to alter a table in the system_auth keyspace, for example, you (as expected) receive the following error:
```'Error from server: code=2100 [Unauthorized] message="<table system_auth.role_members> is protected"'```

However, after I tried to alter each of the tables in those keyspaces (by removing one of the columns), I was able to alter all of the tables in two system keyspaces: `system_distributed` and `system_traces`.

Is this behavior expected? If not, shouldn't an attempt to alter `system_distributed` and `system_traces` result in error 2100 as well?

**Comment by ShlomiBalalis** [comment]
@nyh Is this an expected behavior?

**Comment by dekimir** [comment]
We generally prohibit only alterations that can damage a node irreparably.  For instance, you can't drop any tracing tables.

But we try to allow alterations that can be reversed, since admins use them sometimes to address urgent production issues.

**Comment by fruch** [comment]
> We generally prohibit only alterations that can damage a node irreparably. For instance, you can't drop any tracing tables.
> 
> But we try to allow alterations that can be reversed, since admins use them sometimes to address urgent production issues.

I'm sure I'm following why `system_auth` is different in regards of be open to handle urgent issue.

Do you suggest that if I'll alter the schema of `system_traces` or `system_distributed`, it won't effect scylla behavior at all ?  
than I would like to see what would happen when trying todo trace queries after `system_traces` schema was changed.

**Comment by dekimir** [comment]
`system_auth` is different because if you mess it up you can't even log into Scylla to revert the changes.  Let me double-check what the documented/expected behaviour is for `system_distributed` and `system_traces`...

**Comment by dekimir** [comment]
> Commit [0143aaa](https://github.com/scylladb/scylla/commit/0143aaa5a8f180b76dbb0dbdc50844bba2bf0899) was suppose to stop users from altering the internal (system) keyspaces.

No, it was not.  That commit prevents internal _queries_ from changing the schema of a non-local table.  It's not directly concerned with internal keyspaces.

> When trying to alter a table in the system_auth keyspace, for example, you (as expected) receive the following error:
> `'Error from server: code=2100 [Unauthorized] message="<table system_auth.role_members> is protected"'`

That error was there since Scylla 2.2, and it applies only to `system_auth` tables.  It's not related to 0143aaa.

> However, after I tried to alter each of the tables in those keyspaces (by removing one of the columns), I was able to alter all of the tables in two system keyspaces: `system_distributed` and `system_traces`.
> 
> Is this behavior expected? If not, shouldn't an attempt to alter `system_distributed` and `system_traces` result in error 2100 as well?

I don't know if it's expected (still checking), but the `system_auth` protection is different as explained in my comment above.

**Comment by slivne** [comment]
The users should not have access to change DDL in system tables.

In the past we did change system_traces schema and the upgrade path was for the user to run an alter on system_traces to move it to the new schema.

If we block DDL statements on system_* we need to make sure upgrade code will align the schema during upgrade.

**Comment by dekimir** [comment]
> The users should not have access to change DDL in system tables.

But _superusers_ should, according to #7057.
